dace/transformation/interstate/gpu_transform_sdfg.py

```diff
@@ -192,27 +192,196 @@ def apply(self, _, sdfg: sd.SDFG):
                                                                                    dst_array.desc(sdfg)))
 
         #######################################################
-        # Step 4: Modify transient data storage
+        # Step 4: Change all top-level maps and library nodes to GPU schedule
+
+        gpu_nodes = set()
+        for state in sdfg.nodes():
+            sdict = state.scope_dict()
+            for node in state.nodes():
+                if sdict[node] is None:
+                    if isinstance(node, (nodes.LibraryNode, nodes.NestedSDFG)):
+                        node.schedule = dtypes.ScheduleType.GPU_Default
+                        gpu_nodes.add((state, node))
+                    elif isinstance(node, nodes.EntryNode):
+                        node.schedule = dtypes.ScheduleType.GPU_Device
+                        gpu_nodes.add((state, node))
+                elif self.sequential_innermaps:
+                    if isinstance(node, (nodes.EntryNode, nodes.LibraryNode)):
+                        node.schedule = dtypes.ScheduleType.Sequential
+                    elif isinstance(node, nodes.NestedSDFG):
+                        for nnode, _ in node.sdfg.all_nodes_recursive():
+                            if isinstance(nnode, (nodes.EntryNode, nodes.LibraryNode)):
+                                nnode.schedule = dtypes.ScheduleType.Sequential
+
+        # NOTE: The outputs of LibraryNodes, NestedSDFGs and Map that have GPU schedule must be moved to GPU memory.
+        # TODO: Also use GPU-shared and GPU-register memory when appropriate.
+        for state, node in gpu_nodes:
+            if isinstance(node, (nodes.LibraryNode, nodes.NestedSDFG)):
+                for e in state.out_edges(node):
+                    dst = state.memlet_path(e)[-1].dst
+                    if isinstance(dst, nodes.AccessNode):
+                        desc = sdfg.arrays[dst.data]
+                        desc.storage = dtypes.StorageType.GPU_Global
+            if isinstance(node, nodes.EntryNode):
+                for e in state.out_edges(state.exit_node(node)):
+                    dst = state.memlet_path(e)[-1].dst
+                    if isinstance(dst, nodes.AccessNode):
+                        desc = sdfg.arrays[dst.data]
+                        desc.storage = dtypes.StorageType.GPU_Global
+
+        #######################################################
+        # Step 5: Collect free tasklets and check for scalars that have to be moved to the GPU
+
+        gpu_storage = [dtypes.StorageType.GPU_Global, dtypes.StorageType.GPU_Shared, dtypes.StorageType.CPU_Pinned]
+
+        def _recursive_out_check(node, state, gpu_scalars):
```

Comment:
why not move all three functions out of this method? also why not reuse the recursive version of is_devicelevel_gpu from the other file?